### ISSUE TITLE
Linux OS tuning for InfluxDB

### ISSUE THREAD

**Issue by lesam** [influxdata/influxdb#23246]
This is an epic for different system tuning parameters we have noticed in real installations, and making InfluxDB easier to configure correctly.

Mitigations for bad parameters could include documentation, or InfluxDB doing a self test and logging / stopping / failing to start when misconfigured.

Mitigations for Enterprise and OSS 2 should be considered.

- [ ] vm.max_map_count - if the number of active `mmap` calls is too high, InfluxDB fails to open shards and can even mark database files as corrupt / unqueryable (`.tsm.bad` files will be generated in the data directory). See https://unix.stackexchange.com/questions/640233/vm-max-map-count-for-a-running-process and https://www.elastic.co/guide/en/elasticsearch/reference/current/vm-max-map-count.html for prior art
- [ ] vm.min_free_kbytes - based on internal experimentation, we recommend setting vm.min_free_kbytes to 4.5% of total RAM on machines with more than 50GB of RAM. Compare https://docs.aerospike.com/operations/install/linux/bestpractices . Setting too high can cause OS instability (see https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/performance_tuning_guide/sect-red_hat_enterprise_linux-performance_tuning_guide-configuration_tools-configuring_system_memory_capacity). Setting too low on machines with fast disks and large amounts of RAM appears to cause the OS to trigger the OOM killer even when InfluxDB is using much less than the total available RAM.
- [ ] `ulimit -a` - TSI indexes use many open files (TODO: get a good estimate for open files per shard with TSI indexes). So databases with very large numbers of shards may refuse to start when converted from inmem indexing to TSI indexing. Note that InfluxDB OSS 2 only supports TSI. there are limits on files, number of mmaps, and virtual memory sizes - most of these should be set quite high for InfluxDB.

**Comment by docmerlin** [comment]
Places to check
```bash
cat /proc/{influxd-pid}/limits

cat /proc/sys/fs/nr_open

```
Some notes:
If they need lower default limits, prlimit can be used to increase the limit of a specific pid.

`prlimit -n500000 -p your-pid-here`

`fs.file-max` should be increased to around half a million or so.

**Comment by lesam** [comment]
complete summary of things to check when mmap goes bad:

First, check `free -h` to see if `influxd` is actually using too much memory.

`mmap` is used to both map a file into memory (in order to query it), and to allocate 'anonymous' memory (normal memory not backed by a file). InfluxDB does both of these extensively.

For a full list of errors that the mmap system call on linux can experience, see: https://man7.org/linux/man-pages/man2/mmap.2.html#ERRORS

Many of these are applicable only to file-backed mappings.

For non-file-backed mappings, generally the map will happen in the go runtime, via `sysAlloc`: https://go.dev/src/runtime/mem_linux.go#:~:text=func-,sysAlloc,-(n%20uintptr%2C%20sysStat . This will exit for two common errors, `_EACCESS` (unlikely) and `_EAGAIN` (indicates one should look at `ulimit -l`).

Other errors would result in a `nil` returned from `sysAlloc`, often caught by the go runtime with code similar to:

```
result = (*gcBitsArena)(sysAlloc(gcBitsChunkBytes, &memstats.gcMiscSys))
if result == nil {
	throw("runtime: cannot allocate memory")
}
```

Other errors could be caused by:

`ENFILE`:
Check `cat /proc/sys/fs/file-nr` - this gives 3 numbers. The first is the total file handles on the system, and the last is the maximum. If close to a maximum, consider increasing the limit.

Check `cat /proc/$(pidof influxd)/limits` for file handle limits, and `ls /proc/$(pidof influxd)/fd | wc -l` for the total number of file handles used by `influxd`. If close to a maximum, consider increasing the limit.

`ENOMEM`:
Check `cat /proc/sys/vm/max_map_count` for the per-process limit on memory maps, and `cat /proc/$(pidof influxd)/maps | wc -l` for the total number of maps used by influxd. Increasing this is fairly safe, [SUSE has some docs on side effects](https://www.suse.com/support/kb/doc/?id=000016692) - basically 128 bytes per memory map, and some extra lookups when resolving a memory address. There is no performance penalty unless the maps are actually used.

`EINVAL`:
This would indicate a bug where an invalid size or type of memory region was requested.

**Comment by lesam** [comment]
https://github.com/influxdata/influxdb/issues/23439 - specific issue for adding warnings about these to the logs.

**Comment by LarsStegman** [comment]
We ran into the problem of ` vm.max_map_count` being too low for our dataset. 

The primary server had to be restarted due to a memory leak in a secondary application which caused the machine to start killing processes. 

This would also mark the TSM files as bad, even though this was not the case. If we renamed them from `*.tsm.bad` back to `*.tsm` and opened them in a second influx process, the data would show.

We got errors in the logs while restarting the service indicating: `failed cannot allocate memory`. 

The influx process would crash during startup while reading in our 68TB database. It would crash around 63.8TB of virtual memory being used. Interestingly enough, our back up server which was running at the same time did not encounter this problem during run and is using more than 67TB of virtual memory.

We worked around the problem by increasing the `vm.max_map_count` limit from `65530` to `262144`. The process now starts normally. We suspect this will have increased the limit to 256TB of max virtual memory per process.


We started to discover where the issue was coming from through this issue: https://github.com/prometheus/prometheus/issues/4392
